app/src/main/java/com/rajatgoyal/puzzle15/ui/GameActivity.java

```diff
@@ -513,62 +523,70 @@ public void onClick(DialogInterface dialog, int which) {
-        int i1 = i, j1 = j;
-
-        if (DIR.equals(SWIPE.RIGHT) && isSwipeValid(i, j + 1)) {
+        /*
+          Check if the 4 button around the empty cell is valid or not corresponding to the
+          swipe direction
+         */
+        if (DIR.equals(SWIPE.RIGHT) && isSwipeValid(emptyRowIndex, emptyColIndex - 1)) {
             Timber.d("Swiped Right");
-            j++;
-        } else if (DIR.equals(SWIPE.LEFT) && isSwipeValid(i, j - 1)) {
+            i = emptyRowIndex;
+            j = emptyColIndex - 1;
+        } else if (DIR.equals(SWIPE.LEFT) && isSwipeValid(emptyRowIndex, emptyColIndex + 1)) {
             Timber.d("Swiped Left");
-            j--;
-        } else if (DIR.equals(SWIPE.BOTTOM) && isSwipeValid(i + 1, j)) {
+            i = emptyRowIndex;
+            j = emptyColIndex + 1;
+        } else if (DIR.equals(SWIPE.BOTTOM) && isSwipeValid(emptyRowIndex - 1, emptyColIndex)) {
             Timber.d("Swiped Bottom");
-            i++;
-        } else if (DIR.equals(SWIPE.TOP) && isSwipeValid(i - 1, j)) {
+            i = emptyRowIndex - 1;
+            j = emptyColIndex;
+        } else if (DIR.equals(SWIPE.TOP) && isSwipeValid(emptyRowIndex + 1, emptyColIndex)) {
             Timber.d("Swiped Top");
-            i--;
+            i = emptyRowIndex + 1;
+            j = emptyColIndex;
         } else {
             // Invalid move
             return;
         }
 
         updateMoves(++moves);
+
+        // Todo: Make this click sound optional, give option from the setting to turn this on or off
         playClickSound();
 
         // swapping of tiles
-        m[i][j] = num;
-        buttons[i][j].setText(String.format("%s", num));
-        buttons[i][j].setBackgroundColor(getResources().getColor(R.color.background));
+        int num = Integer.parseInt(buttons[i][j].getText().toString());
+        m[emptyRowIndex][emptyColIndex] = num;
+        buttons[emptyRowIndex][emptyColIndex].setText(String.format("%s", num));
+        buttons[emptyRowIndex][emptyColIndex].setBackgroundColor(getResources().getColor(R.color.background));
 
-        m[i1][j1] = 0;
-        buttons[i1][j1].setText("");
-        buttons[i1][j1].setBackgroundColor(getResources().getColor(R.color.light));
+        m[i][j] = 0;
+        buttons[i][j].setText("");
+        buttons[i][j].setBackgroundColor(getResources().getColor(R.color.light));
+
+        emptyRowIndex = i;
+        emptyColIndex = j;
 
         if (checkIfGameOver()) {
             wonGame();
         }
     }
 
-    private boolean isSwipeValid(int i, int j) {
+    /**
+     * @param rowIndex row index of the surrounding buttons w.r.t to the empty cell
+     * @param colIndex column index of the surrounding buttons w.r.t to the empty cell
+     * @return if there exists a valid button around the empty cell corresponding to the
+     * swipe direction, returns true if it exists, false otherwise
+     */
+    private boolean isSwipeValid(int rowIndex, int colIndex) {
```

Comment:
since swipe changes only one of the coordinates, no need to check both coordinates.. just make a function `isIndexOutOfBound` or something like this and use it in `isMoveValid` too

Reply:
Done in 97916924f16e00d0530397dfcaf672ec7cd12bd5